pageserver/src/tenant/timeline.rs

```diff
@@ -788,6 +743,22 @@ impl Timeline {
         result
     }
 
+    pub(super) fn spawn_flush_loop(self: &Arc<Self>) {
```

Comment:
Let's add a boolean to Timeline struct, to keep track of whether the flush loop has already been started. If we start two flush loops on the same timeline, bad things will happen, and it's currently not very clear when it's started or who's responsible for starting it. Would be good to make that more clear, but a boolean would go a long way. We could then just not start it again if its already running, which would allow you to just always call spawn_flush_loop() if it might or might not be running. In particular, `initialize_with_lock` could just always call spawn_flush_loop(), it wouldn't need the new `spawn_flush_loop` argument.

Reply:
Yeah, flush_loop supervision is one of my concerns. It is possible for it to stop due to panic (not that high of a probability, but still) and in this case checkpointer wont be able to make progress. I thought about adding bool to timeline to avoid double start problem but decided against it since it was only two cases and in other ones flag is equal to load_layer_map one we already have there. Additionally `initialize` without flags is also used for "default" cases. So I moved manual invocations to flush_loop and call to checkpoint to the same function so it is easier to see such places. I agree that having a bool would be safer. Will change. I also think that double start should be considered as an error (debug_assert maybe?)

Ideally we would rather store a join handle to check if task is alive and restart it when it isnt. Or add this logic to the task manager (aka restart_on_error flag to task_mgr::spawn)

Reply:
Hmm, remembered another reason why I did it this way. If we start flush_loop only in `initialize()` then checkpoint needs to happen after that, so we register timeline as ready and observable by `get_timeline` calls before we actually finish the whole import routine (is it correct to treat checkpoint as a part of this process?). We also start listening for wal_receiver events.

Is it a real concern or its ok to do it either way?

Reply:
Settled with flag option. @hlinnaka ping, if it looks ok to you lets merge it :)